cmd/kubeadm/app/cmd/token.go

```diff
@@ -0,0 +1,183 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"text/tabwriter"
+	"time"
+
+	"github.com/spf13/cobra"
+
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	kubemaster "k8s.io/kubernetes/cmd/kubeadm/app/master"
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/kubectl"
+)
+
+const (
+	bootstrapTokenSecretPrefix = "bootstrap-token-"
+	defaultTokenDuration       = time.Duration(8) * time.Hour
+)
+
+func NewCmdToken(out io.Writer) *cobra.Command {
+	tokenCmd := &cobra.Command{
+		Use:   "token",
+		Short: "Manage bootstrap tokens",
+	}
+
+	var tokenDuration time.Duration
+	tokenSecret := &kubeadmapi.Secrets{}
+	createCmd := &cobra.Command{
+		Use:   "create",
+		Short: "Create discovery tokens on the server.",
+		Run: func(tokenCmd *cobra.Command, args []string) {
+			err := RunCreateToken(out, tokenCmd, tokenDuration, tokenSecret)
+			kubeadmutil.CheckErr(err)
+		},
+	}
+	createCmd.PersistentFlags().DurationVar(&tokenDuration,
+		"duration", defaultTokenDuration, "Duration the token should be valid")
+	createCmd.PersistentFlags().StringVar(
+		&tokenSecret.GivenToken, "token", "",
+		"Shared secret used to secure cluster bootstrap; if none is provided, one will be generated for you",
+	)
+	tokenCmd.AddCommand(createCmd)
+
+	listCmd := &cobra.Command{
+		Use:   "list",
+		Short: "List discovery tokens on the server.",
+		Run: func(tokenCmd *cobra.Command, args []string) {
+			err := RunListTokens(out, tokenCmd)
+			kubeadmutil.CheckErr(err)
+		},
+	}
+	tokenCmd.AddCommand(listCmd)
+
+	deleteCmd := &cobra.Command{
+		Use:   "delete",
+		Short: "Delete discovery tokens on the server.",
+		Run: func(tokenCmd *cobra.Command, args []string) {
+			err := RunDeleteToken(out, tokenCmd, args[0])
+			kubeadmutil.CheckErr(err)
+		},
+	}
+	tokenCmd.AddCommand(deleteCmd)
+
+	return tokenCmd
+}
+
+// RunCreateToken generates a new bootstrap token and stores it as a secret on the server.
+func RunCreateToken(out io.Writer, cmd *cobra.Command, tokenDuration time.Duration, tokenSecret *kubeadmapi.Secrets) error {
+	client, err := kubemaster.CreateClientFromFile(path.Join(kubeadmapi.GlobalEnvParams.KubernetesDir, "admin.conf"))
+	if err != nil {
+		return err
+	}
+
+	err = kubeadmutil.GenerateTokenIfNeeded(tokenSecret)
+	if err != nil {
+		return err
+	}
+
+	secret := &api.Secret{
+		ObjectMeta: api.ObjectMeta{
+			Name: fmt.Sprintf("%s%s", bootstrapTokenSecretPrefix, tokenSecret.TokenID),
+		},
+		Type: api.SecretTypeBootstrapToken,
+		Data: encodeTokenSecretData(tokenSecret, tokenDuration),
+	}
+	if _, err := client.Secrets(api.NamespaceSystem).Create(secret); err != nil {
```

Comment:
We should make this an `upsert` where we update or create/insert.

Do this:
1. Run in a loop with a certain number of retries in case something flakes out.  Something like 5.
2. Try to get the secret.  
  1. If it is there, then replace the Data array and update.
  2. If it doesn't exist (`IsNotFound`) then create it and set it
3. If you get an error then retry from the top.  Errors might be `IsNotFound` (if you are doing an update and it was deleted between the get/update) or `IsConflict` (if you raced someone to do an update and lost).


See `pkg/api/errors/errors.go`